src/System.Buffers.Primitives/System/Buffers/OwnedBuffer.cs

```diff
@@ -41,7 +39,7 @@ public bool HasOutstandingReferences
         {
             get
             {
-                return _referenceCount > 0
+                return Volatile.Read(ref _referenceCount) > 0
```

Comment:
I'm not sure, but I added this because I'm worried about the Debug.Assert(_owned.HasOutstandingReferences) in BufferSegment.Dispose().

Isn't possible that without this volatile read, a BufferSegment could be created and then disposed on a different thread that sees a cashed value of _owned._referenceCount from before the BufferSegment called _owned.AddReference() in the ctor?